lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3397,45 +3810,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-                                                    field);
+    SILValue pullback =
+        builder.createStructExtract(loc, primalValueAggregateInAdj, field);
 
-    // Construct the pullback arguments.
+    // Get the original result of the `apply` instruction.
     SmallVector<SILValue, 8> args;
-    auto av = takeAdjointValue(ai);
-    ValueWithCleanup seedBuf(builder.createAllocStack(loc, av.getType()),
+    SmallVector<SILValue, 8> origDirResults;
+    collectAllExtractedElements(ai, origDirResults);
+    SmallVector<SILValue, 8> origAllResults;
+    collectAllActualResultsInTypeOrder(
+        ai, origDirResults, ai->getIndirectSILResults(),
+        origAllResults);
+    auto origResult = origAllResults[applyInfo.actualIndices.source];
+    auto origNumIndRes = ai->getNumIndirectResults();
+
+    auto pullbackType =
+        remapType(pullback->getType()).castTo<SILFunctionType>();
+    SILFunctionConventions pullbackConvs(pullbackType, getModule());
+
+    // Get adjoint value of the original result.
+    SILType adjointType;
+    SILValue adjointValue;
+    if (origResult->getType().isObject()) {
+      // If original result is a `tuple_extract`, materialize adjoint value of
+      // `ai` and extract the corresponding element adjoint value.
+      if (auto *tupleExtract = dyn_cast<TupleExtractInst>(origResult)) {
+        auto adjointTuple = materializeAdjoint(takeAdjointValue(ai), loc);
+        adjointValue = builder.emitTupleExtract(loc, adjointTuple,
+                                                tupleExtract->getFieldNo());
+      }
+      // Otherwise, materialize adjoint value of `ai`.
+      else {
+        adjointValue = materializeAdjoint(takeAdjointValue(origResult), loc);
+      }
+      adjointType = adjointValue->getType();
+    } else {
+      auto adjointBuf = getAdjointBuffer(origResult);
+      adjointValue = adjointBuf.getValue();
+      adjointType = adjointBuf.getValue()->getType();
+    }
+
+    // Register indirect results as arguments.
+    // FIXME: This logic is crucial and needs to be fixed.
+    SILFunctionConventions calleeConvs(
+        ai->getCallee()->getType().castTo<SILFunctionType>(), ai->getModule());
+    SmallVector<SILValue, 4> tempAllocations;
+    unsigned count = pullbackType->getNumIndirectFormalResults();
+    for (auto paramIdx : range(calleeConvs.getNumParameters())) {
+      if (count == 0)
+        break;
+      auto paramInfo = calleeConvs.getParameters()[paramIdx];
+      if (paramInfo.isFormalIndirect()) {
+        auto param = ai->getArgument(
+            calleeConvs.getNumIndirectSILResults() + paramIdx);
+        auto *buf = builder.createAllocStack(loc,
+            remapType(getCotangentType(param->getType(), getModule())));
+        tempAllocations.push_back(buf);
+        args.push_back(buf);
+        ++count;
+      }
+    }
+
+    // Initialize seed buffer.
+    ValueWithCleanup seedBuf(builder.createAllocStack(loc, adjointType),
```

Comment:
I don't think `ValueWithCleanup` is necessary here. It's not being used by any calls below that take `ValueWithCleanup`.

Reply:
Changed `seedBuf` to just `SILValue` in https://github.com/apple/swift/pull/22544/commits/c1ff15405abb1df3ec85d58a0b40df718376a90f.